File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {TreeNode}
     */
    invertTreeBFS(root: TreeNode | null): TreeNode {
        if (!root) return root;
        let current = root;
        let quee = [current];
        let front = 0;

        while (front < quee.length) {
            let node = quee[front++];
            if (node.left) quee.push(node.left);
            if (node.right) quee.push(node.right);
            let temp = node.right;
            node.right = node.left;
            node.left = temp;
        }
        return root;
    }
    invertTree(root: TreeNode | null): TreeNode {
        if (!root) return root;
        let temp = root.right;
        root.right = root.left;
        root.left = temp;
        if (root.left) this.invertTree(root.left);
        if (root.right) this.invertTree(root.right);
        return root;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def check(node, min=float("-inf"), max=float("inf")):
            if not node:
                return True
            if node.val <= min or node.val >= max:
                return False
            return check(node.left, min, node.val) and check(node.right, node.val, max)

        return check(root)
